### PR TITLE
Implement strings API in test executor

### LINKED ISSUE #198: Support strings api in test executor
The [test executor](https://github.com/zio/zio-redis/blob/master/redis/src/main/scala/zio/redis/Interpreter.scala#L80) currently only supports a subset of the api.
The goal of this issue is to support the full strings api in the test executor. Tests for this group should be enabled for the test executor [here](https://github.com/zio/zio-redis/blob/master/redis/src/test/scala/zio/redis/ApiSpec.scala#L32)

### PR DESCRIPTION
**Pull request by mvelimir**
Resolves #198 
[BITFIELD](https://redis.io/commands/bitfield) not yet implemented

### CODE REVIEW

**Comment by jdegoes** [comment]
Looks really great! Would be good to have some tests for the new operations to make sure they are working as specified.

**Comment by mijicd** [comment]
@mvelimir Please resolve the conflicts.

**Comment by mijicd** [comment]
@mvelimir Is this still work in progress? If not, please convert it to a pull request :)

**Comment by mvelimir** [comment]
There's still the BITFIELD command left to be implemented, but I could make a separate PR for that.

**Comment by mijicd** [comment]
@mvelimir I agree, let's have it in a separate PR. Please rebase this one, and I'll review it immediately.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 1]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+            strings.getOrElse(key, "").map { string =>
+              Some(bit).filter(bit => bit == "0" || bit == "1").fold[RespValue](Replies.Error) { bit =>
+                Try(start.getOrElse("0").toInt).toOption.fold[RespValue](Replies.Error) { start =>
+                  Try(end.getOrElse(string.length.toString).toInt).toOption.fold[RespValue](Replies.Error) { end =>
+                    val newStart = if (start < 0) string.length + start else start
+                    val newEnd   = (if (end < 0) string.length + end else end) + 1
+
+                    val index = string.zipWithIndex
+                      .slice(newStart, newEnd)
+                      .map(tuple => f"${tuple._1.asInstanceOf[Int].toBinaryString}%8s".replace(' ', '0') -> tuple._2)
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3007]: Let's get rid of `asInstanceOf`.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 2]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+
+        orMissingParameter2(keyOption, decrOption) { (key, decr) =>
+          orWrongType(isString(key))(
+            strings.get(key).flatMap { stringOption =>
+              Try(stringOption.getOrElse("0").toLong).toOption.fold(STM.succeed(Replies.Error: RespValue)) { num =>
+                if (decr >= 0 && num >= Long.MinValue + decr || decr < 0 && num <= Long.MaxValue + decr) {
+                  val result = num - decr
+
+                  putString(key, result.toString).as(RespValue.Integer(result))
+                } else STM.succeed(Replies.Error)
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3053]: Simplify as suggested above.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 3]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+        orMissingParameter(keyOption) { key =>
+          orWrongType(isString(key))(
+            strings.get(key).flatMap { stringOption =>
+              Try(stringOption.getOrElse("0").toLong).toOption.fold(STM.succeed(Replies.Error: RespValue)) { num =>
+                if (num > Long.MinValue) {
+                  val result = num - 1
+
+                  putString(key, result.toString).as(RespValue.Integer(result))
+                } else STM.succeed(Replies.Error)
+              }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3036]: Simplify as suggested above.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 4]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+        val offsetOption = input.lift(1).map(_.asString)
+
+        orMissingParameter2(keyOption, offsetOption) { (key, offset) =>
+          orWrongType(isString(key))(
+            strings.getOrElse(key, "").map { string =>
+              Try(offset.toInt).toOption.filter(_ >= 0).fold[RespValue](Replies.Error) { offset =>
+                RespValue
+                  .Integer(string.getBytes.lift(offset / 8).fold(0L)(byte => (byte >> (7 - offset % 8) & 1).toLong))
+              }
+            }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3070]: Ditto.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 5]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+            strings.getOrElse(key, "").map { string =>
+              Try(start.toInt).toOption
+                .flatMap(a => Try(end.toInt).toOption.map(b => (a, b)))
+                .fold[RespValue](Replies.Error) { range =>
+                  val newStart = if (range._1 < 0) string.length + range._1 else range._1
+                  val newEnd   = (if (range._2 < 0) string.length + range._2 else range._2) + 1
+
+                  RespValue.bulkString(string.slice(newStart, newEnd))
+                }
+            }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3090]: Ditto

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 6]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+
+        orMissingParameter(keyOption) { key =>
+          orWrongType(isString(key))(
+            strings.get(key).flatMap { stringOption =>
+              Try(stringOption.getOrElse("0").toLong).toOption.fold(STM.succeed(Replies.Error: RespValue)) { num =>
+                if (num < Long.MaxValue) {
+                  val result = num + 1
+
+                  putString(key, result.toString).as(RespValue.Integer(result))
+                } else STM.succeed(Replies.Error)
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3118]: You can use `ZSTM.cond`

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 7]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+            strings.getOrElse(key, "0").flatMap { string =>
+              Try(string.toLong).toOption.fold(STM.succeed(Replies.Error: RespValue)) { num =>
+                Try(incr.toLong).toOption.fold(STM.succeed(Replies.Error: RespValue)) { incr =>
+                  if (incr >= 0 && num <= Long.MaxValue - incr || incr < 0 && num >= Long.MinValue - incr) {
+                    val result = num + incr
+
+                    putString(key, result.toString).as(RespValue.Integer(result))
+                  } else STM.succeed(Replies.Error)
+                }
+              }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3139]: Simplify as suggested above.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 8]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+          orWrongType(isString(key))(
+            strings.getOrElse(key, "0").flatMap { string =>
+              Try(string.toDouble).toOption.fold(STM.succeed(Replies.Error: RespValue)) { num =>
+                Try(incr.toDouble).toOption.fold(STM.succeed(Replies.Error: RespValue)) { incr =>
+                  if (incr >= 0 && num <= Double.MaxValue - incr || incr < 0 && num >= Double.MinValue - incr) {
+                    val result = (num + incr).toString
+
+                    putString(key, result).as(RespValue.bulkString(result))
+                  } else STM.succeed(Replies.Error)
+                }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3158]: Simplify as suggested above.

File: redis/src/main/scala/zio/redis/TestExecutor.scala [hunk 9]
```diff
@@ -2886,6 +2905,696 @@ private[redis] final class TestExecutor private (
...
+
+                    newString.lift(offset / 8).fold(STM.succeed(Replies.Error: RespValue)) { char =>
+                      val resp = (char >> (7 - offset % 8) & 1).toLong
+                      val updatedChar = (if (value == 1) char | (1 << (7 - offset % 8))
+                                         else char & ~(1 << (7 - offset % 8))).toChar
+
+                      putString(key, newString.updated(offset / 8, updatedChar)).as(RespValue.Integer(resp))
+                    }
+                  }
+              }
```
Comment (mijicd) [inline redis/src/main/scala/zio/redis/TestExecutor.scala:3342]: Simplify as suggested above.

**Review by mijicd — CHANGES_REQUESTED** [review]
First pass done.

**Comment by jdegoes** [comment]
@mvelimir Would love to get this in! Do you have a chance to push it through?

**Comment by mvelimir** [comment]
Sorry for the wait.

**Review by mijicd — APPROVED** [review]

**Pull request merged**